include/openmc/mesh.h

```diff
@@ -155,59 +185,54 @@ class StructuredMesh : public Mesh {
   //
   //! \param[in] Array of mesh indices
   //! \return Mesh bin
-  virtual int get_bin_from_indices(const int* ijk) const;
+  virtual int get_bin_from_indices(const MeshIndex& ijk) const;
 
   //! Get mesh indices given a position
   //
   //! \param[in] r Position to get indices for
-  //! \param[out] ijk Array of mesh indices
   //! \param[out] in_mesh Whether position is in mesh
-  virtual void get_indices(Position r, int* ijk, bool* in_mesh) const;
+  //! \return Array of mesh indices
+  virtual MeshIndex get_indices(Position r, bool& in_mesh) const;
 
   //! Get mesh indices corresponding to a mesh bin
   //
   //! \param[in] bin Mesh bin
-  //! \param[out] ijk Mesh indices
-  virtual void get_indices_from_bin(int bin, int* ijk) const;
+  //! \return ijk Mesh indices
+  virtual MeshIndex get_indices_from_bin(int bin) const;
 
   //! Get mesh index in a particular direction
   //!
   //! \param[in] r Coordinate to get index for
   //! \param[in] i Direction index
   virtual int get_index_in_direction(double r, int i) const = 0;
 
-  //! Check where a line segment intersects the mesh and if it intersects at all
-  //
-  //! \param[in,out] r0 In: starting position, out: intersection point
-  //! \param[in] r1 Ending position
-  //! \param[out] ijk Indices of the mesh bin containing the intersection point
-  //! \return Whether the line segment connecting r0 and r1 intersects mesh
-  virtual bool intersects(Position& r0, Position r1, int* ijk) const;
-
-  //! Get the coordinate for the mesh grid boundary in the positive direction
-  //!
-  //! \param[in] ijk Array of mesh indices
-  //! \param[in] i Direction index
-  virtual double positive_grid_boundary(int* ijk, int i) const = 0;
-
-  //! Get the coordinate for the mesh grid boundary in the negative direction
+  //! Get the closest distance from the coordinate r to the grid surface  
+  //! in i direction  that bounds mesh cell ijk and that is larger than l
+  //! The coordinate r does not have to be inside the mesh cell ijk. In 
+  //! curved coordinates, multiple crossings of the same surface can happen,
+  //! these are selected by the parameter l 
   //!
   //! \param[in] ijk Array of mesh indices
-  //! \param[in] i Direction index
-  virtual double negative_grid_boundary(int* ijk, int i) const = 0;
+  //! \param[in] i direction index of grid surface
+  //! \param[in] r0 position, from where to calculate the distance
+  //! \param[in] u direction of flight. actual position is r0 + l * u
+  //! \param[in] l actual chord length
+  //! \return MeshDistance struct with closest distance, next cell index in i-direction and min/max surface indicator
+  virtual MeshDistance distance_to_grid_boundary(const MeshIndex& ijk, int i, const Position& r0, const Direction& u, double l) const = 0;
```

Comment:
This parameter is required to find the right solution in circular coordinates! But I can change this by updating the `r0` position which results in an implicit `l=0`. This is a better design, as is reduced the redundant update. But I will check, if this could introduce rounding errors.

Reply:
Nice! Yeah I think that will also make the algorithm in `raytrace_mesh` a little more clear.

Reply:
I check the source. It is possible, but will result in a (very small) performance degradation, especially in rectilinear grids:
- to get rid of the `l` parameter, the `r` need to be `r+l*u`
- The distances to the next grid surface are always measured from `r0` in order to have an absolute scale and be able to sort these.
- For rectilinear grids, these distance can be simply calculated as `r0/u` without even multiplying `r+l*u`
- As i'm only interested in the distance in one grid direction, (e.g. x) the calculation of y and z components are totally useless for rectilinear grids.